### ISSUE TITLE
Two requests get 401, but only one request is send again

### ISSUE THREAD

**Issue by israelKusayev** [Flyrell/axios-auth-refresh#56]
I send two requests at once, but after the refresh token only the last request get called again

![image](https://user-images.githubusercontent.com/42112796/70985669-27681480-20c5-11ea-9310-a45edd265053.png)

**Comment by Flyrell** [comment]
Can you please provide the minimal reproduction of the issue?

**Comment by israelKusayev** [comment]
**Version:** 2.0.2

**Configurations:**
```javascript
import createAuthRefreshInterceptor from 'axios-auth-refresh';


// Add a request interceptor
axios.interceptors.request.use(function (config) {
    config.headers.Authorization = 'Bearer ' + authService.getAccessToken();
    return config;
});

// Function that will be called to refresh authorization
const refreshAuthLogic = (failedRequest: any) =>
  authService.refreshToken(authService.getRefreshToken()!)
    .then(res => {
      authService.storeTokens(res.data);
      return Promise.resolve();
    })
    .catch(() => {
      authService.logout();
      history.push('/login');
      return Promise.reject();
    });

// Instantiate the interceptor (you can chain it as it returns the axios instance)
createAuthRefreshInterceptor(axios, refreshAuthLogic);
```

**Requests**

<details><summary>First GET request</summary>
<img src="https://user-images.githubusercontent.com/42112796/71083951-d4fa2700-219c-11ea-8954-55f63bf11297.png"> </img>
</details>


<details><summary>Second GET request</summary>
<img src="https://user-images.githubusercontent.com/42112796/71084198-6bc6e380-219d-11ea-92bc-fae7d3584167.png"> </img>
</details>


<details><summary>Rrefresh token request</summary>
requesrt
<img src="https://user-images.githubusercontent.com/42112796/71084381-c95b3000-219d-11ea-8120-f3484f501f37.png"> </img>
response
<img src="https://user-images.githubusercontent.com/42112796/71084432-eee83980-219d-11ea-83ff-f240135300bb.png"> </img>
</details>

After the token is refreshed, only the second request gets called again

<details><summary>second GET request again</summary>
<img src="(https://user-images.githubusercontent.com/42112796/71084913-f9570300-219e-11ea-8aa8-dccc103c1530.png"> </img>
</details>


<details><summary>Chrome requests summery</summary>
<img src="(https://user-images.githubusercontent.com/42112796/71084995-2b686500-219f-11ea-8936-5c6b16435a27.png"> </img>
</details>

**Comment by Flyrell** [comment]
Right, I think I know what is causing the issue here. Can you confirm, that the request which responds first with 401 status is called again? So in the case you provided it should be the `Second GET request` that responded faster than the `First Get request`. If so, the following should be true...

This is caused because of the instance is marked as `paused` inside the interceptor, as the `skipAuthInterceptor` feature was not working correctly in releases `1.x`. Now it works, so it should be fairly ease to fix. Logically this is what happens step-by-step:

1. `Request #1` is called
2. `Request #2` is called
3. `Request #2` responds with 401 status
4. **Response interceptor is triggered** by the `Request #2`
5. **Instance is marked as paused** inside the **Response interceptor** to prevent the loop (there was not skipAuthInterceptor option when this was implemented)
6. **Request interceptor is created** by the **Response interceptor**
6.1. *From now on, all the requests are stalled until the refreshing is done*
7. **Refreshing logic is called** by the **Response interceptor**
8. `Request #1` responds with 401 status
9. **Response interceptor skips the response** as axios instance is marked as `paused` and the refreshing logic did not complete yet.
...

It will require deletion of the whole "paused instances" logic as it is not needed now as we have `skipAuthInterceptor` option available. Thanks for clarifying the issue.

**Comment by israelKusayev** [comment]
Yes, I made a mistake.
The `first get request` is called again (which returns first 401) not the second request.

Are you planning to delete the "paused instances" logic soon?

**Comment by Flyrell** [comment]
@israelKusayev I'll try to take a look at it tomorrow, maybe a day after.

**Comment by keyvanm** [comment]
Hi! Does setting the `{ skipWhileRefreshing: false }` option mitigate this bug for now with no side effects? It seems to fix the issue, though I want to make sure there are no breaking side effects. The only thing I see is that the refresh logic is called twice, which puts extra load on the server, but that is fine.

**Comment by israelKusayev** [comment]
It solves the problem,    and the refresh logic isn't called twice!

<details><summary>The requests...</summary>
<img src="https://user-images.githubusercontent.com/42112796/71444998-ac36dc00-271e-11ea-8122-4b452d6226c6.png"> </img>
</details>

Is it the right way to solve this problem?

**Comment by Flyrell** [comment]
Sorry, things did not exactly go as I planned them to and I literally have no time to invest in this issue now.

Of course, I will get to it ASAP. If `skipWhileRefreshing` works for you, I'm glad I thought about it back then when I was implementing the paused instances, but I feel like it can be a little bit buggy when the refresh logic call itself returns 401.

In case anyone wants to help with the fix mentioned above, I can of course do a Code Review and Merge the changes it if there are no problems. 

> It will require deletion of the whole "paused instances" logic as it is not needed now as we have skipAuthInterceptor option available.

**Comment by Flyrell** [comment]
Ok, now I see what I meant by creating the whole `skipWhileRefreshing` logic...

If the `skipWhileRefreshing` is disabled, you need to implement the "loop prevention" yourself. As mentioned in [here](https://github.com/Flyrell/axios-auth-refresh#unpause-the-instance-while-refreshing) it will allow the program to enter the interceptor loop which means that whenever the refresh logic returns any code from [`statusCodes`](https://github.com/Flyrell/axios-auth-refresh#status-codes-to-intercept) option (default `[ 401 ]`) it will intercept it once again and you'd enter the loop.

Skip instances wouldn't be needed if the library would have access to your request object, but the refresh logic can be whatever you want, not only an API call, which makes it a bit tricky. To prevent not-advanced users from entering the loop, the paused instances were added. 

**TL;DR**
The plugin will work as expected, but you need to disable the [instance pausing](https://github.com/Flyrell/axios-auth-refresh#unpause-the-instance-while-refreshing) using the `skipWhileRefreshing` as you mentioned before and add the `{ skipAuthRefresh: true }` to your refreshing call in order to prevent the loops.

**Comment by Flyrell** [comment]
@israelKusayev @keyvanm can you please confirm that what I wrote previously works for you? Bear in mind that `skipAuthRefresh` option will work only with axios v0.18 and then in v0.19.1 (which is not published yet), so the current v0.19 does not work as there is a bug that strips custom properties from the request object.

**Comment by israelKusayev** [comment]
`{ skipWhileRefreshing: false }` seems to solve the problem
I had problems with Axios 0.19.0 (when the refresh token is invalid the logout logic didn't work)
I downgraded Axios to 0.18.0 and it works OK

**Comment by israelKusayev** [comment]
so `{ skipWhileRefreshing: false }` solves this bug?

**Comment by Flyrell** [comment]
Yes, well... it's not a bug actually. The library just doesn't have a control over the call in your `authRefresh` function, which is why it pauses the whole instance in case the call causes the 401 (which would otherwise turn into an infinite loop). It might be worth to think it over once again, but for now it works that way. Just don't forget to set the `{ skipAuthRefresh: true }` option on your refreshing call when you use `{ skipWhileRefreshing: false }` option.

**Comment by keyvanm** [comment]
Setting `{ skipWhileRefreshing: false }` solves the problem. They only side effect was that occasionally the refresh logic was called twice, which I think setting `{ skipAuthRefresh: true }` will resolve.

**Comment by ankitjaiswal1994** [comment]
This has been handled in the latest v3.0.0.
Thanks @Flyrell
